File: P904 Fruit Into Baskets /Solution.java
class Solution {
    public int totalFruit(int[] tree) {
        Set<Integer> types = new HashSet();
        int maxCount = 0, tempCount = 0, i = 0, j = 0;
        while(i < tree.length) {
            int n = tree[i];
            types.add(n);
            if(types.size() == 3) {
                if(maxCount < tempCount) {
                    maxCount = tempCount;
                }
                tempCount = 0;
                j = i;
                types.clear();
                while(types.size() < 2) {
                    j-=1;
                    types.add(tree[j]);
                }
                i = j+1;
                types.clear();
                continue;
            } else {
                tempCount++;
            }
            i++;
        }
        if(maxCount < tempCount) {
            maxCount = tempCount;
        }
        return maxCount;
    }
}

File: P904 Fruit Into Baskets /solution.py
class Solution:
    def totalFruit(self, tree: List[int]) -> int:
        types  = []
        # j goes backward
        max_count = temp_count = i = j = 0
        while i < len(tree):
            n = tree[i]
            if n in types:
                temp_count += 1
            else:
                if len(types) == 2:
                    if max_count < temp_count:
                        max_count = temp_count
                    temp_count = 0
                    j = i
                    types = []
                    while len(types) < 2:
                        j-=1
                        if tree[j] not in types:
                            types.append(tree[j])
                    i = j+1
                    types = []
                    continue
                else:
                    types.append(n)
                    temp_count += 1
            i+=1
                    
        if max_count < temp_count:
            max_count = temp_count
                    
        return max_count